### PR TITLE
[aotinductor] reland: return a copy of any constant

### PR DESCRIPTION
**Pull request by chenyang78**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112370

When the model returns a constant, we cannot "release" its handle,
because the constant doesn't have any handle at all. Instead,
we should allocate a new tensor and then return a copy of the constant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by hl475** [comment]
Thanks @chenyang78 ! Could you please import this PR to internal to collect more signals? IIRC the original PR got reverted due to internal tests breakage?

**Comment by chenyang78** [comment]
> Thanks @chenyang78 ! Could you please import this PR to internal to collect more signals? IIRC the original PR got reverted due to internal tests breakage?

Thanks, @hl475. I ran the aot_inductor_test internally and it passed. Looks like we cannot import to fbsource. Let me make a temporary diff based on this PR to trigger internal CI.

**Review by hl475 — APPROVED** [review]
LG!

**Review by desertfire — APPROVED** [review]

**Comment by swolchok** [comment]
is this sufficient? isn't it possible to alias a constant via codegen_exact_buffer_reuse?

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -1499,25 +1499,44 @@ def define_kernel(
...
+                    if config.aot_inductor.abi_compatible:
                         self.wrapper_call.writeline(
-                            f"aoti_torch_new_uninitialized_tensor(&output_handles[{idx}]);"
+                            f"aoti_torch_clone({output}, &output_handles[{idx}]);"
                         )
+                    else:
                         self.wrapper_call.writeline(
-                            f"aoti_torch_assign_tensors({output}, output_handles[{idx}]);"
+                            f"output_handles[{idx}] = reinterpret_cast<AtenTensorHandle>("
+                            + f"new at::Tensor(std::move({output}.clone())));"
```
Comment (swolchok) [inline torch/_inductor/codegen/wrapper.py:1519]: the std::move here is not necessary; the result of .clone() is already a prvalue . fixing in conflict resolution with #112116

**Comment by chenyang78** [comment]
> is this sufficient? isn't it possible to alias a constant via codegen_exact_buffer_reuse?

Yeah, I think it's sufficient, because we don't reuse constant buffers:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/wrapper.py#L1147

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (chenyang78) [reply torch/_inductor/codegen/wrapper.py:1519]: Good call. Thanks for fixing it!

**Pull request closed without merging**